### PR TITLE
Swap with a route (more than one liquidity pool)

### LINKED ISSUE #19: [bindings-test] Support swaps with a non-empty route
Currently, the mocks in `osmo-bindings-test` only support swaps with an empty route. [They panic otherwise.](https://github.com/confio/osmosis-bindings/blob/7c020283cb190a8768a5f635599f1d5ed6f9d7a9/packages/bindings-test/src/multitest.rs#L225-L232)

This means the only swaps we can test are direct ones with one liquidity pool.

```rust
// pool 1 is ("OSMO", "ATOM")
// OSMO -> ATOM swap
let swap = OsmosisMsg::Swap {
    first: Swap {
        pool_id: 1,
        denom_in: "OSMO".to_string(),
        denom_out: "ATOM".to_string(),
    },
    route: vec![],
    amount: ...,
};
```

We cannot have more complex ones that go through multiple liquidity pools:

```rust
// pool 1 is ("OSMO", "ATOM")
// pool 2 is ("ATOM", "BTC")
// OSMO -> ATOM -> BTC swap
let swap = OsmosisMsg::Swap {
    first: Swap {
        pool_id: 1,
        denom_in: "OSMO".to_string(),
        denom_out: "ATOM".to_string(),
    },
    route: vec![Step {
        pool_id: 2,
        denom_out: "BTC".to_string(),
    }],
    amount: ...,
};
```

Without this, we cannot properly test https://github.com/confio/isotonic/issues/92 or https://github.com/confio/isotonic/issues/101. Let's implement this.

### PR DESCRIPTION
**Pull request by uint**
Closes #19 

Swaps can now have a route and still be tested with `bindings-test`!

### CODE REVIEW

File: packages/bindings-test/src/multitest.rs [hunk 1]
```diff
@@ -540,4 +592,128 @@ mod tests {
         ];
         assert_eq!(state.assets, expected_assets);
     }
+
+    #[test]
+    fn perform_swap_with_route_exact_out() {
+        let pool1 = Pool::new(coin(6_000_000, "osmo"), coin(3_000_000, "atom"));
+        let pool2 = Pool::new(coin(2_000_000, "atom"), coin(1_000_000, "btc"));
```
Comment (maurolacy) [inline packages/bindings-test/src/multitest.rs:599]: I want access to this pool.

File: packages/bindings-test/src/multitest.rs [hunk 2]
```diff
@@ -540,4 +592,128 @@ mod tests {
...
+                pool_id: 1,
+                denom_in: "osmo".to_string(),
+                denom_out: "atom".to_string(),
+            },
+            route: vec![Step {
+                pool_id: 2,
+                denom_out: "btc".to_string(),
+            }],
+            amount: SwapAmountWithLimit::ExactIn {
+                input: Uint128::new(4028),
```
Comment (maurolacy) [inline packages/bindings-test/src/multitest.rs:687]: Why not use a round value here (4000) and check fees are proper.

File: packages/bindings-test/src/multitest.rs [hunk 3]
```diff
@@ -540,4 +592,128 @@ mod tests {
...
+        };
+        let res = app.execute(trader.clone(), msg.into()).unwrap();
+
+        let Coin { amount, .. } = app.wrap().query_balance(&trader, "osmo").unwrap();
+        assert_eq!(amount, Uint128::new(5000 - 4033));
+        let Coin { amount, .. } = app.wrap().query_balance(&trader, "btc").unwrap();
+        assert_eq!(amount, Uint128::new(1000));
+
+        // check the response contains proper value
+        let input: EstimatePriceResponse = from_slice(res.data.unwrap().as_slice()).unwrap();
```
Comment (maurolacy) [inline packages/bindings-test/src/multitest.rs:637]: This should be called `SwapResponse`. Changing it is kind of difficult now, as it'll impact osmosis.

**Review by maurolacy — APPROVED** [review]
LGTM.

Tests in which `max_input` / `min_output` are exceeded would be good, just to confirm error behaviour.

Same with testing pools with the wrong denoms.

File: packages/bindings-test/src/multitest.rs (hunk 1 above)
Comment (uint) [reply packages/bindings-test/src/multitest.rs:599]: Who knew you can still get rich on BTC?

**Comment by uint** [comment]
Forgot about `EstimateSwap`, on it...

**Comment by uint** [comment]
Got a query implementation in. Refactoring the code was a surprising PITA. Tests and polish coming on Monday at this point!

File: packages/bindings/src/types.rs [hunk 4]
```diff
@@ -64,3 +64,12 @@ pub enum SwapAmountWithLimit {
     ExactIn { input: Uint128, min_output: Uint128 },
     ExactOut { output: Uint128, max_input: Uint128 },
 }
+
+impl SwapAmountWithLimit {
+    pub fn discard_limit(self) -> SwapAmount {
```
Comment (maurolacy) [inline packages/bindings/src/types.rs:69]: 👍🏼

**Review by maurolacy — APPROVED** [review]
Looks good. Tests for non-happy paths are still missing. I think this is pretty good for being test code already, so approving (again).

**Pull request merged**